File: backend/databaseUrl.js
"use strict";

const POSTGRES_COMPONENT_KEYS = [
    "POSTGRES_HOST",
    "POSTGRES_PORT",
    "POSTGRES_USER",
    "POSTGRES_PASSWORD",
    "POSTGRES_DB",
];

/**
 * Resolve the PostgreSQL connection URL without changing an explicit URL.
 *
 * @param {Record<string, string | undefined>} environment process environment
 * @returns {string | undefined} explicit, constructed, or absent database URL
 */
function resolveDatabaseUrl(environment) {
    const explicitDatabaseUrl = environment.DATABASE_URL;
    if (explicitDatabaseUrl) return explicitDatabaseUrl;

    const hasAllComponents = POSTGRES_COMPONENT_KEYS.every(
        (key) => environment[key],
    );
    if (!hasAllComponents) return explicitDatabaseUrl;

    const user = encodeURIComponent(environment.POSTGRES_USER);
    const password = encodeURIComponent(environment.POSTGRES_PASSWORD);
    return (
        `postgresql://${user}:${password}@${environment.POSTGRES_HOST}:` +
        `${environment.POSTGRES_PORT}/${environment.POSTGRES_DB}`
    );
}

module.exports = { resolveDatabaseUrl };

File: services/common/database_url.py
"""PostgreSQL connection URL resolution shared by Python sidecars."""

from collections.abc import Mapping
from urllib.parse import quote


def resolve_database_url(environment: Mapping[str, str]) -> str:
    """Preserve an explicit URL or build one from complete PostgreSQL components."""
    explicit_database_url = environment.get("DATABASE_URL", "")
    if explicit_database_url:
        return explicit_database_url

    host = environment.get("POSTGRES_HOST")
    port = environment.get("POSTGRES_PORT")
    user = environment.get("POSTGRES_USER")
    password = environment.get("POSTGRES_PASSWORD")
    database = environment.get("POSTGRES_DB")
    if not host or not port or not user or not password or not database:
        return explicit_database_url

    encoded_user = quote(user, safe="")
    encoded_password = quote(password, safe="")
    return f"postgresql://{encoded_user}:{encoded_password}@{host}:{port}/{database}"
